lib/spack/spack/config.py

```diff
@@ -639,13 +664,28 @@ def they_are(t):
 
     # Source dict is merged into dest.
     elif they_are(dict):
+        # track keys for marking
+        key_marks = {}
+
         for sk, sv in iteritems(source):
             if _override(sk) or sk not in dest:
                 # if sk ended with ::, or if it's new, completely override
                 dest[sk] = copy.copy(sv)
             else:
                 # otherwise, merge the YAML
                 dest[sk] = _merge_yaml(dest[sk], source[sk])
+
+            # this seems unintuitive, but see below. We need this because
+            # Python dicts do not overwrite keys on insert, and we want
+            # to copy mark information on source keys to dest.
+            key_marks[sk] = sk
+
+        # ensure that keys are marked in the destination.  the key_marks dict
+        # ensures we can get the actual source key objects from dest keys
+        for dk in dest.keys():
+            if dk in key_marks:
+                syaml.mark(dk, key_marks[dk])
```

Comment:
Super minor point (because this is already a huge a leap forward what we have right now): a key of a dictionary that is effectively a merge of more files is attributed to one of them. Example:
```console
$ spack config blame packages
---                                                                     packages:
/home/mculpo/PycharmProjects/spack/etc/spack/packages.yaml:2              all:
/home/mculpo/PycharmProjects/spack/etc/spack/defaults/packages.yaml:18      compiler:
/home/mculpo/PycharmProjects/spack/etc/spack/defaults/packages.yaml:18      - gcc
/home/mculpo/PycharmProjects/spack/etc/spack/defaults/packages.yaml:18      - intel
/home/mculpo/PycharmProjects/spack/etc/spack/defaults/packages.yaml:18      - pgi
/home/mculpo/PycharmProjects/spack/etc/spack/defaults/packages.yaml:18      - clang
/home/mculpo/PycharmProjects/spack/etc/spack/defaults/packages.yaml:18      - xl
/home/mculpo/PycharmProjects/spack/etc/spack/defaults/packages.yaml:18      - nag
/home/mculpo/PycharmProjects/spack/etc/spack/packages.yaml:3                providers:
<omitted>
/home/mculpo/PycharmProjects/spack/etc/spack/defaults/packages.yaml:25        gl:
/home/mculpo/PycharmProjects/spack/etc/spack/packages.yaml:5                  - mesa
/home/mculpo/PycharmProjects/spack/etc/spack/defaults/packages.yaml:26        glu:
/home/mculpo/PycharmProjects/spack/etc/spack/defaults/packages.yaml:26        - mesa-glu
/home/mculpo/PycharmProjects/spack/etc/spack/defaults/packages.yaml:26        - openglu
<omitted>
```
Above I think that `all` and `providers` should either have no tag, or have a tag that says they are the result of a merge.

Reply:
Yeah, this is a reasonable point.  I guess it's somewhat obvious to me that it's a merge if the values have multi-file annotations.

As I don't have a quick fix for this, I think I will leave it for a future PR.